Move/Fix FakeTensor logic for detecting multiple fake modes

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97186

This was leftover for when we had more logic in the FakeTensor and not FakeTensorMode, and wasn't firing correctly. It also makes more sense for it to be in the other validation function.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

eellison committed:
Move/Fix FakeTensor logic for detecting multiple fake modes

[ghstack-poisoned]

bdhirsh reviewed (approved):
this is great, would have saved me some debugging pain a while ago

eellison committed:
Update on "Move/Fix FakeTensor logic for detecting multiple fake modes"


This was leftover for when we had more logic in the FakeTensor and not FakeTensorMode, and wasn't firing correctly. It also makes more sense for it to be in the other validation function.

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/passes/decomp.py:
```diff
@@ -48,7 +48,7 @@ def graph_with_interpreter(*args):
     decomposed_module = proxy_tensor.make_fx(
         graph_with_interpreter,
         decomposition_table=decomposition_table,
-        tracing_mode="fake",
```
cc @wschin

eellison committed:
Update on "Move/Fix FakeTensor logic for detecting multiple fake modes"


This was leftover for when we had more logic in the FakeTensor and not FakeTensorMode, and wasn't firing correctly. It also makes more sense for it to be in the other validation function.

[ghstack-poisoned]

eellison committed:
Update on "Move/Fix FakeTensor logic for detecting multiple fake modes"


This was leftover for when we had more logic in the FakeTensor and not FakeTensorMode, and wasn't firing correctly. It also makes more sense for it to be in the other validation function.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

eellison commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -657,8 +657,11 @@ def disable_autocast_cache():
 
-def make_fx(f, decomposition_table=None, tracing_mode="real", _allow_non_fake_inputs=False):
-    assert tracing_mode in ["real", "fake", "symbolic"]
+def make_fx(f, decomposition_table=None, tracing_mode="real", _allow_non_fake_inputs=False, existing_fake_mode=None):
```
cc @Chillee @bdhirsh @voznesenskym @ezyang 

i made these api changes. feel free to bikeshed.

BowenBao commented on torch/onnx/_internal/fx/passes/shape_inference.py:
```diff
@@ -18,7 +18,12 @@ def shape_inference_with_fake_tensor(decomposed_module: torch.fx.GraphModule, *a
     # If (1) and (2) are done with difference FakeTensorMode's, undefined behavior may
     # happen.
-    fake_tensor_mode = fake_tensor.FakeTensorMode()
+    fake_tensor_mode = torch._dynamo.utils.fake_mode_from_tensors(args)  # type: ignore[arg-type]
```
nit
```suggestion
    fake_tensor_mode = torch._dynamo.utils.fake_mode_from_tensors(args) or fake_tensor.FakeTensorMode()
```

wschin replied:
I remember we want to use the `init_env` in FakeTensorMode to override some shapes. If we can't create FakeTensorMode here, then we need to modify FakeTensorMode's API to add a flag. @titaiwangms, for heads up.

wschin commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -657,8 +657,11 @@ def disable_autocast_cache():
-def make_fx(f, decomposition_table=None, tracing_mode="real", _allow_non_fake_inputs=False):
-    assert tracing_mode in ["real", "fake", "symbolic"]
+def make_fx(f, decomposition_table=None, tracing_mode="real", _allow_non_fake_inputs=False, existing_fake_mode=None):
+    assert tracing_mode in ["real", "fake", "symbolic", "existing"]
```
`fake` and `existing` looks duplicated. When caller specifying `existing`, I assume the code will create FakeTensorMode if it doesn't exist. Since there is no document, I don't quite understand the design concepts and expected behaviors of those options.

eellison committed:
Update on "Move/Fix FakeTensor logic for detecting multiple fake modes"


This was leftover for when we had more logic in the FakeTensor and not FakeTensorMode, and wasn't firing correctly. It also makes more sense for it to be in the other validation function.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

eellison committed:
Update on "Move/Fix FakeTensor logic for detecting multiple fake modes"


This was leftover for when we had more logic in the FakeTensor and not FakeTensorMode, and wasn't firing correctly. It also makes more sense for it to be in the other validation function.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

eellison committed:
Update on "Move/Fix FakeTensor logic for detecting multiple fake modes"


This was leftover for when we had more logic in the FakeTensor and not FakeTensorMode, and wasn't firing correctly. It also makes more sense for it to be in the other validation function.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

eellison commented on torch/onnx/_internal/fx/passes/decomp.py:
```diff
@@ -32,6 +37,30 @@ def _rename_placeholder_targets(
     module.recompile()
 
 
+def wrapped_fn(fn, mode, allow_non_fake_inputs, args):
```
@bdhirsh @ezyang this is "wrapping" of a callable so that users can trace with an existing fake mode. IMO, it is a lot to expect a user to write, especially considering the number of times we've seen a user attempt to do something like this themselves.... however i'm really just trying to unblock the fake tensor error so I will do whatever to land this

ezyang replied:
I agree we need to work this out a better way to do this. Maybe we can discuss this tomorrow at composability social. My current thinking is that if we can enforce there is one fake mode, it makes sense to have an easy context manager `with_fake` and `no_fake`.

eellison commented:
cc @wschin @BowenBao I'm trying to make a change that will error when multiple FakeTensorModes are active, since this is not supported and leads to confusing errors currently. On the pr there were onnx failures because it would call `make_fx` with FakeTensor inputs, so there would exist both the mode from the inputs and the newly constructed fake tensor mode. I have implemented a wrapper so that there would not be multiple fake modes around. Now i'm getting other onnx errors I dont know how to interpret/fix. Would you please take a look ?

BowenBao commented on torch/onnx/_internal/fx/passes/decomp.py:
```diff
@@ -32,6 +37,30 @@ def _rename_placeholder_targets(
+        )
+
+    args = [
+        (mode.from_tensor(t)
```
@eellison onnx error from this line. `mode` is a str it appears?

BowenBao replied:
guess should be `fake_mode`?

eellison replied:
Thanks!

BowenBao replied:
@wschin for `export_without_parameters_and_buffers`

wschin replied:
I don't have clear idea. I need to see how the `input_` passed into that `assert`. It might be related to argument-wrapping code in exporter. If I get sometime this week, I will try debug.

wschin replied:
I am building this branch. Let's see if I can reproduce it tomorrow.

BowenBao commented on torch/onnx/_internal/fx/passes/decomp.py:
```diff
@@ -32,6 +37,30 @@ def _rename_placeholder_targets(
     module.recompile()
 
 
+def wrapped_fn(fn, mode, allow_non_fake_inputs, args):
```
Related to onnx error comment. Type annotating the arguments helps bring a lot clarity.

eellison committed:
Update on "Move/Fix FakeTensor logic for detecting multiple fake modes"


This was leftover for when we had more logic in the FakeTensor and not FakeTensorMode, and wasn't firing correctly. It also makes more sense for it to be in the other validation function.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wschin commented on torch/onnx/_internal/fx/passes/decomp.py:
```diff
@@ -69,14 +98,18 @@ def graph_with_interpreter(*args):
+            fn,
             decomposition_table=self.decomposition_table,
-            tracing_mode=fx_mode,
+            tracing_mode="real",
```
`real`? I guess we always want to run decomposition with `fake` or `symbolic`. Maybe you can add another flag `reuse_existing_fake_or_create_fake`.?

eellison replied:
That was the flag I tried to add which people didn't want me to lol. There may be changes in the future to the api after this pr.

wschin replied:
Humm, I like that flag. If we change the value to `real`, exporter will be way slower.

eellison replied:
we're not actually doing real, we're still wrapping the inputs to fake. We're avoiding calling `real` so we dont do a new FakeTensorMode on top of the mode from the inputs.

wschin replied:
What if the input must be a FakeTensor?

In the error above, the whole model is initialized under FakeTensorMode, so all its parameters/buffers are FakeTensor's already. I guess that's why you see that assert. I am not sure how to deal with it without running the code. Building PyTorch on my side will take a while.

eellison replied:
need any pointers ?

eellison commented:
cc @bowenbao @wschin can you please a take a look ?

BowenBao commented on torch/onnx/_internal/fx/passes/decomp.py:
```diff
@@ -32,6 +37,30 @@ def _rename_placeholder_targets(
+    context = enable_python_dispatcher if mode == "symbolic" else contextlib.nullcontext
+    shape_env = ShapeEnv() if mode == "symbolic" else None
+    fake_mode = torch._dynamo.utils.fake_mode_from_tensors(args)
+    if mode is None:
```
```suggestion
    if fake_mode is None:
```
I think this resolves the onnx errors in CI. I'm not sure about the one you posted though, building locally to have a look.

BowenBao commented:
@eellison with new changes I'm getting all test cases failing under the test file, even for ones that were not creating their own fake tensor mode. I'm not sure how to deal with those.
```
---  TestFxToOnnxWithOnnxRuntime.test_simple_function ---
...
  File "/home/bowbao/pytorch/torch/fx/experimental/proxy_tensor.py", line 99, in get_proxy_slot
    raise RuntimeError(f"{obj} is not tracked with proxy for {tracer}")
RuntimeError: s0 is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7f5423302190>

While executing %add : [#users=2] = call_function[target=operator.add](args = (%x, 1.0), kwargs = {})
Original traceback:
  File "/home/bowbao/pytorch/test/onnx/test_fx_to_onnx_with_onnxruntime.py", line 110, in func
    y = x + 1.0
```

For the other one that creates its own fake tensor mode outside, I guess the solution needs to be extended to look for fake tensor mode from model weights as well.

BowenBao commented:
@eellison so the problem is with how "user code" stash up the contexts outside. The current code below doesn't work because it does not accurately resemble what happens inside `make_fx`.
```python
    def fn_wrapped():
        with context():
            with fake_mode:
                return fn(*args)
```
Instead, it should be something like
```python
        fake_args, shape_env, fake_mode, context = wrapped_fn(
            graph_with_interpreter, fx_mode, allow_non_fake_inputs=True, args=args
        )

        with fake_mode, context():
            decomposed_module = proxy_tensor.make_fx(
                graph_with_interpreter,
                decomposition_table=self.decomposition_table,
                tracing_mode="real",
                _allow_non_fake_inputs=True,
            )(*fake_args)
        if fx_mode == "symbolic":
            decomposed_module.shape_env = shape_env
```

Frankly the issue had nothing to do with onnx. With this change I'm passing the onnx tests. tbh this looks very messy, I'm looking forward to the general fake_mode follow up. cc @ezyang

ezyang commented:
At composability sync, the tentative agreement was that we would make make_fx, by default, automatically detect if you pass in fake inputs or have fake mode and use that directly; or if there was no fake mode automatically fakeify.

ezyang commented:
What's holding up landing this?

BowenBao commented:
@eellison could you try update the PR with my suggestions above? That should unblock the onnx CI failures.

eellison committed:
Update on "Move/Fix FakeTensor logic for detecting multiple fake modes"


This was leftover for when we had more logic in the FakeTensor and not FakeTensorMode, and wasn't firing correctly. It also makes more sense for it to be in the other validation function.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

eellison committed:
Update on "Move/Fix FakeTensor logic for detecting multiple fake modes"


This was leftover for when we had more logic in the FakeTensor and not FakeTensorMode, and wasn't firing correctly. It also makes more sense for it to be in the other validation function.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

eellison commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -698,15 +698,23 @@ def wrapped(*args):
-            fake_tensor_mode = FakeTensorMode(
-                allow_fallback_kernels=True,
-                allow_non_fake_inputs=_allow_non_fake_inputs)
+            fake_tensor_mode = torch._dynamo.utils.detect_fake_mode(args)
```
how do these changes look @ezyang ?

ezyang replied:
Yep, looks great!

eellison commented:
@pytorchbot merge

PaliC commented:
@pytorchbot revert

huydhn commented:
@pytorchbot revert -m 'This breaks ONNX test in trunk and it looks like a landrace as the CI signal is green' -c landrace

BowenBao commented:
@eellison do you have a link to the failing onnx ci log?

BowenBao commented:
Error is from `torch/onnx/_internal/fx/passes/shape_inference.py` where a new `fake_tensor_mode` was created and used before calling `FakeTensorPropGetStaticShapes(module, fake_tensor_mode).propagate(*args)`. Looks like now the proper way should be  try grabbing existing mode by `fake_tensor_mode = torch._dynamo.utils.detect_fake_mode(args)` before creating the new mode. Is that correct @eellison ?

cc @titaiwangms

eellison commented:
@BowenBao yes that's correct

BowenBao commented:
Rebase with master and apply this diff should unblock.

```diff
diff --git a/torch/onnx/_internal/fx/passes/shape_inference.py b/torch/onnx/_internal/fx/passes/shape_inference.py
index a708f354509..14a41a8f277 100644
--- a/torch/onnx/_internal/fx/passes/shape_inference.py
+++ b/torch/onnx/_internal/fx/passes/shape_inference.py
@@ -4,6 +4,7 @@ import itertools
 from typing import Optional
 
 import torch
+import torch._dynamo
 import torch.fx
 from torch._subclasses import fake_tensor
 from torch.fx.experimental import proxy_tensor
@@ -33,7 +34,9 @@ class ShapeInferenceWithFakeTensor(_pass.Transform):
         # 2. run FakeTensorProp
         # If (1) and (2) are done with difference FakeTensorMode's, undefined behavior may
         # happen.
-        fake_tensor_mode = fake_tensor.FakeTensorMode()
+        fake_tensor_mode = (
+            torch._dynamo.utils.detect_fake_mode(args) or fake_tensor.FakeTensorMode()
+        )
 
         def to_fake_tensor(x):
             if isinstance(x, torch.Tensor) and not isinstance(
```

eellison committed:
Update on "Move/Fix FakeTensor logic for detecting multiple fake modes"


This was leftover for when we had more logic in the FakeTensor and not FakeTensorMode, and wasn't firing correctly. It also makes more sense for it to be in the other validation function.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge
